app/controllers/newflow/login_signup_controller.rb

```diff
@@ -0,0 +1,323 @@
+    def oauth_callback
+      handle_with(
+        OauthCallback,
+        success: lambda {
+          authentication = @handler_result.outputs.authentication
+          user = @handler_result.outputs.user
+          unless user.is_activated?
+            # not activated means signup
+            save_unverified_user(user)
+            @first_name = user.first_name
+            @last_name = user.last_name
+            @email = @handler_result.outputs.email
+            security_log(:student_social_sign_up, user: user, authentication_id: authentication.id)
+            # must confirm their social info on signup
+            render :confirm_social_info_form and return
+          end
+          sign_in!(user)
+          # we should perhaps create a more specific security log
+          security_log(:student_authenticated_with_social, user: user, authentication_id: authentication.id)
+          redirect_back(fallback_location: profile_newflow_path)
+        },
+        failure: lambda {
+          # TODO: rate-limit this
+          @email = @handler_result.outputs.email
+          save_login_failed_email(@email)
+          security_log(:student_auth_with_social_failed)
+          render :social_login_failed
+        }
+      )
+    end
+
+    def confirm_oauth_info
+      handle_with(
+        ConfirmOauthInfo,
+        user: unverified_user,
+        contracts_required: !contracts_not_required,
+        client_app: get_client_app,
+        success: lambda {
+          clear_newflow_state
+          sign_in!(@handler_result.outputs.user)
+          security_log(:student_social_auth_confirmation_success)
+          redirect_to signup_done_path
+        },
+        failure: lambda {
+          security_log(:student_social_auth_confirmation_failed)
+          render :confirm_social_info_form
+        }
+      )
+    end
+
+    def setup_password
+      handle_with(
+        SetPassword,
+        user: current_user,
+        success: lambda {
+          security_log(:student_setup_password)
+          redirect_back(fallback_location: profile_newflow_url)
+        },
+        failure: lambda {
+          security_log(:student_setup_password_failed, { user: @handler_result.outputs.user })
+          render :setup_password_form
+        }
+      )
+    end
+
+  private #################
+
+    def known_signup_role_redirect
+      known_role = session.fetch(:signup_role, nil)
+
+      if known_role && known_role == 'student'
+        redirect_to newflow_signup_student_path(request.query_parameters) # TODO: when we create the Educator flow, redirect to there.
+      end
+    end
+
+    def cache_client_app
+      set_client_app(params[:client_id])
+    end
+
+    def save_unverified_user(user)
+      session[:unverified_user_id] = user.id
+    end
+
+    def unverified_user
+      id = session[:unverified_user_id]&.to_i
+      return unless id.present?
+
+      @unverified_user ||= User.find_by(id: id, state: 'unverified')
```

Comment:
They don't cleaned up somewhere, currently. But I don't think it's actually a problem because unverified users represent only one specific kind of user: users who started but have not finished the signup process. Reports that don't want to count that type of users should/could easily omit them in a `where.not` clause.

We _could_ clean up users who have not finished the sign up process in more than [too long to care anymore] number of days — in a rake task.

Reply:
I created a card for this: https://github.com/openstax/business-intel/issues/865